### DISCUSSION TITLE
Plans for the Core Data dependency

### DISCUSSION THREAD

**Discussion by KaiOelfke** [tgrapperon/swift-dependencies-additions#34] [Q&A]
The Core Data dependency code is interesting for me, because I use Ensembles in some projects. I'm curious what the plan is here. I've started to play around with the Core Data Case Study mostly to better understand the code. I've pushed it, if that's interesting for you or others. I will continue in the next days or so.

https://github.com/KaiOelfke/swift-dependencies-additions/tree/core-data-experiments

There's still many things that are not quite clear for me.

For example I'd like to see how this can work with relationships and what it takes to add and remove composers for example in the Add Song Form.

I understand that this dependency is experimental. So there's no expectations from me but I just wanted to highlight my interest.

Thank you for your work @tgrapperon

**Comment by tgrapperon** [comment]
Hello @KaiOelfke! Thanks for the experiment and for opening the discussion!

I think that one objective for the Core Data dependency would be to reimplement a `FetchRequest` property wrapper similar to the one that ships with SwiftUI, like we did for `AppStorage`. I have promising drafts in this direction, but I lacked the time to polish them. 

I would also like the dependency to harden thread safety and make it easier to work with managed objects. SwiftUI has a simpler problem because most interactions happen on the main thread, and their CoreData entities are living in the `viewContext`. It's slightly more complex when having to deal with background contexts. I would also like to preserve sync access to `viewContext` entities on the `MainActor`.

I'll try to improve the relationship aspect very soon. It should be able to work with Ensembles if you're able to initialize the `\.persistentContainer` dependency with a `NSPersistentContainer` that is bound to `Ensembles`.

Feel free to ask questions for things that are not clear in this thread!

**Reply by KaiOelfke** [reply]
A `FetchRequest` replacement would be great. Let me know, if I can help you somehow. If you'd like feedback on your drafts I can have a look, if you want.
